File: apps/api/app/modules/analytics/schemas/analytics.py
"""Response DTOs for the analytics module. Plain ``pydantic.BaseModel`` (not
SQLModel) — these are read-only projections, never persisted."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel

from app.modules.raffles.models import RaffleStatus


class ExecutiveDashboard(BaseModel):
    """Top-level KPIs (Sprint 9). Independent from the Sprint-7 operational
    dashboard (`app.modules.dashboard`) — this module runs its own queries."""

    raffles_total: int
    raffles_published: int
    raffles_closed: int
    tickets_available: int
    tickets_reserved: int
    tickets_paid: int
    participants_total: int
    collaborators_total: int
    # "Ingresos esperados": sum(total_tickets * ticket_price) across matched raffles.
    expected_revenue: float
    # "Ingresos recibidos": sum(ticket_price) for PAID tickets. Derived from
    # ticket status, not a real transaction — the seat reserved for the future
    # Payments module (see docs/06-roadmap/ROADMAP.md).
    received_revenue: float
    percent_sold: float
    percent_reserved: float


class RankingEntry(BaseModel):
    """One row of a Top-N ranking (collaborator, raffle, or participant)."""

    id: uuid.UUID
    name: str
    count: int
    value: float
    rank: int


class RaffleReportRow(BaseModel):
    id: uuid.UUID
    title: str
    status: RaffleStatus
    created_at: datetime
    draw_date: datetime
    total_tickets: int
    available: int
    reserved: int
    paid: int
    percent_sold: float
    expected_revenue: float


class RaffleReportDetail(RaffleReportRow):
    """Single-raffle report, including the two rankings — only computed for one
    raffle at a time (never per-row in the list) to avoid N+1 queries."""

    top_collaborators: list[RankingEntry]
    top_participants: list[RankingEntry]


class CollaboratorReportRow(BaseModel):
    id: uuid.UUID
    name: str
    raffle_id: uuid.UUID
    raffle_title: str
    total_sales: int
    total_reservations: int
    participants_served: int
    expected_amount: float
    participation_percent: float
    rank: int


class ParticipantReportRow(BaseModel):
    id: uuid.UUID
    full_name: str
    purchases_count: int
    tickets_count: int
    amount_invested: float
    last_purchase_at: datetime | None
    raffles_count: int


class TimeSeriesPoint(BaseModel):
    day: date
    count: int
    value: float


class TicketStatusDistribution(BaseModel):
    available: int
    reserved: int
    paid: int
    cancelled: int
    winner: int


class GlobalReports(BaseModel):
    top_collaborators: list[RankingEntry]
    top_raffles: list[RankingEntry]
    top_participants: list[RankingEntry]
    sales_by_day: list[TimeSeriesPoint]
    reservations_by_day: list[TimeSeriesPoint]
    status_distribution: TicketStatusDistribution

File: packages/api-client/src/dto/analytics.ts
/**
 * Wire-format DTOs for the Analytics & Reports module (Sprint 9).
 * Field names are snake_case on purpose — they mirror the JSON the API sends
 * (apps/api `modules/analytics`), not internal TS naming conventions.
 */
import type { RaffleStatus } from './raffle';
import type { TicketStatus } from './ticket';

export interface ExecutiveDashboard {
  raffles_total: number;
  raffles_published: number;
  raffles_closed: number;
  tickets_available: number;
  tickets_reserved: number;
  tickets_paid: number;
  participants_total: number;
  collaborators_total: number;
  /** "Ingresos esperados": total_tickets * ticket_price across matched raffles. */
  expected_revenue: number;
  /** "Ingresos recibidos": sum of ticket_price for PAID tickets — a seat for
   * the future Payments module, not a real transaction yet. */
  received_revenue: number;
  percent_sold: number;
  percent_reserved: number;
}

export interface RankingEntry {
  id: string;
  name: string;
  count: number;
  value: number;
  rank: number;
}

export interface RaffleReportRow {
  id: string;
  title: string;
  status: RaffleStatus;
  created_at: string;
  draw_date: string;
  total_tickets: number;
  available: number;
  reserved: number;
  paid: number;
  percent_sold: number;
  expected_revenue: number;
}

export interface RaffleReportDetail extends RaffleReportRow {
  top_collaborators: RankingEntry[];
  top_participants: RankingEntry[];
}

export interface CollaboratorReportRow {
  id: string;
  name: string;
  raffle_id: string;
  raffle_title: string;
  total_sales: number;
  total_reservations: number;
  participants_served: number;
  expected_amount: number;
  participation_percent: number;
  rank: number;
}

export interface ParticipantReportRow {
  id: string;
  full_name: string;
  purchases_count: number;
  tickets_count: number;
  amount_invested: number;
  last_purchase_at: string | null;
  raffles_count: number;
}

export interface TimeSeriesPoint {
  day: string;
  count: number;
  value: number;
}

export interface TicketStatusDistribution {
  available: number;
  reserved: number;
  paid: number;
  cancelled: number;
  winner: number;
}

export interface GlobalReports {
  top_collaborators: RankingEntry[];
  top_raffles: RankingEntry[];
  top_participants: RankingEntry[];
  sales_by_day: TimeSeriesPoint[];
  reservations_by_day: TimeSeriesPoint[];
  status_distribution: TicketStatusDistribution;
}

/** Shared filters accepted by every analytics endpoint (see
 * docs/04-api/API_SPECIFICATION.md, "Analytics"). `status` is a ticket status;
 * it does not apply to the raffles report (see `raffle_status` below). */
export interface AnalyticsFiltersQuery {
  [key: string]: string | number | boolean | undefined | null;
  start_date?: string | undefined;
  end_date?: string | undefined;
  raffle_id?: string | undefined;
  status?: TicketStatus | undefined;
  collaborator_id?: string | undefined;
}

export interface AnalyticsRafflesQuery extends AnalyticsFiltersQuery {
  raffle_status?: RaffleStatus | undefined;
  page?: number;
  page_size?: number;
}

export interface AnalyticsCollaboratorsQuery extends AnalyticsFiltersQuery {
  page?: number;
  page_size?: number;
}

export interface AnalyticsParticipantsQuery extends AnalyticsFiltersQuery {
  page?: number;
  page_size?: number;
}

export type AnalyticsExportFormat = 'excel' | 'pdf';
